### ISSUE TITLE
time representation as fractions of seconds

### ISSUE THREAD

**Issue by sigbjorn** [statkraft/shyft#211]
# Overview

Current implementation uses time as seconds utc 1970, int64 (fast and easy).

We need sub-seconds resolution, e.g. ms.

Change internal rep to use ratio based time-rep,  eg.  chrono.

Ensure python exposure understands time  (int) and time(float) as.

**Comment by jfburkhart** [comment]
Hi, we should carry out this discussion here and not on email...  I just want to make one comment, which may not be immediately relevant to #211 but should be on our minds. There are many issues with time and time steps in the routines that make some assumptions regarding the delta-t.

For example: "The ability to separate the hypothesized model equations from their solutions, especially if the latter require numerical approximations. For example, for a continuous-simulation rainfall-runoff model, the governing equations should be formulated and reported in continuous-time state-space form, and only then approximated in discrete time [e.g., Kavetski et al., 2003; Young and Garnier, 2006; Clark and Kavetski, 2010, and references therein]" From: paragraph 27: http://onlinelibrary.wiley.com/doi/10.1029/2010WR009827/full

**Comment by sigbjorn** [comment]
Thanks!, a good point and reminder. I will add the email shortly. It is more related to the concrete implementation of this issue, and the current version of the chrono PR #213.

**Comment by sigbjorn** [comment]
# Review questions, thoughts and comments from email:

Regarding PR #213, tag chrono_v1, that exposes the 'classical' chrono types time_point and duration as UtcTime and TimeSpan to python.
C++ core using time_point and duration types.

The question is whether this kind of non-numeric type-representation for time is suitable in Shyft context.

The chrono_v1 tag implementation clearly demonstrates that it works, and that we are able to both change time-resolution (the initial objective of the issue), as well as enforcing different types for time-point (UtcTime) and duration(TimeSpan).

The choice of UtcTime/TimeSpan was to ease other statkraft internal software that have similar concepts, inherited from .NET space.

My current conclusion is that this split of representation brings in more problems/issues than it was meant to solve. A large portion of the effort in the PR, leading into the chrono_v1 tag was ensuring that UtcTime was exposed as number, and that TimeSpan also was exposed as a number, and that they did work as numbers.

To illustrate the issues found, consider the chrono (and most time/date and calendar libraries approach):

 time_point is not a number, it's an object, where the internal rep is seconds/ticks since epoch.
 time_point * number is not allowed etc.

but:
   time_point  - time_point -> duration , that works almost like a number.
 and
  time_point + duration -> time_point, 

In analytical software, time is mostly viewed as 4th dimension or similar, with it's standard SI-measurement unit [seconds], and represented as number.
This is exactly how most science is done, the type of things are numbers, the variables denotes what number we are talking about.

E.g. distance:
```python
distance_1=  3000.0   # type is number
distance_2 = 2000.0   # type is number
distance_3 = 1000.0   # .. 
distance_delta = distance_2 - distance_1  # result type is also number, no surprise here
avg_distance = (distance_1 + distance_2 + distance_3)/3  # average, no problems, type is number
distance_4 = 2*distance_3 - distance_2 # a number, prediction of next distance based on two last
```
As we can see, the variables and the types are additive information, and it works by symmetry and rules that is rather well established.

The classical time-libraries, where there are separate types for time_point and duration, violates these rules, and claims it's for the best for all of us, because otherwise we could intermix something like a time-point and a duration, which is considered different things. 
- and yes, they are different items, by definition (but still numbers),
  but no, it is not difficult, nor unusual, and we don't need the help of separate types to help us.
 - it's what we have variables for, and adding use of SI-units to calculations make this easy.

The example above do work for any kind, like weight, speed etc. It even work for time ;-).
 
.. unless you introduce time_point and duration (not specific to chrono, it applies to other time, date calendar libaries as well!).

```python
t_1=  UtcTime('2003-10-01T12:30:00Z')   # type is UtcTime, internal rep seconds since 1970utc
t_2=  UtcTime('2003-10-01T12:30:00Z')   # type is UtcTime
t_3=  UtcTime('2003-10-01T12:30:00Z')   # type is UtcTime
t_delta = t_2 -t_1  # result type is TimeSpan!  internal rep also number, seconds
t_avg= (t_1 + t_2 + t_3)/3  # fail! not allowed all ops are invalid
t_4 = 2*t_3 - t_2 # fail! num * UtcTime  not allowed
t_4 = t_3 + (t_3-t_2)  # this is ok! types matches
```

So basically a lot of pain, without contributing with any new vital information.

The only nice thing so far with UtcTime as type, is that it can display it's well defined time as iso 8601 string, similar for the TimeSpan. But this is for debug purposes, and can easiliy be provided by specific number-types as well.

I am aiming for some more work on the chrono PR, where we transform the types into symmetric number types (aliases) that are more suited for scientific and analytical software.
